charmcraft/commands/store/__init__.py

```diff
@@ -1396,26 +1355,50 @@ def run(self, parsed_args):
             resource_type = ResourceType.file
             logger.debug("Uploading resource directly from file %s", resource_filepath)
         elif parsed_args.image:
+            image_digest = parsed_args.image
+            credentials = store.get_oci_registry_credentials(
+                parsed_args.charm_name, parsed_args.resource_name
+            )
+            image_name = credentials.image_name.split("/", 1)[1]
             logger.debug(
-                "Uploading resource from image %s at Dockerhub", parsed_args.image
+                "Uploading resource from image %s @ %s", image_name, image_digest
             )
-            full_image_name = "/".join(
-                (parsed_args.image.organization, parsed_args.image.name)
+
+            # build the image handler
+            registry = OCIRegistry(
+                self.config.charmhub.registry_url,
+                image_name,
+                username=credentials.username,
+                password=credentials.password,
             )
-            registry = OCIRegistry("https://registry.hub.docker.com", full_image_name)
             ih = ImageHandler(registry)
-            final_resource_url = ih.get_destination_url(parsed_args.image.reference)
-            logger.debug("Resource URL: %s", final_resource_url)
-            resource_type = "oci-image"
-
-            # create a JSON pointing to the unique image URL (to be uploaded to Charmhub)
-            resource_metadata = {
-                "ImageName": final_resource_url,
-            }
+
+            # check if the specific image is already in Canonical's registry
+            already_uploaded = ih.check_in_registry(image_digest)
+            if already_uploaded:
+                logger.info("Using OCI image from Canonical's registry")
+            else:
+                # upload it from local registry
+                logger.info("Remote image not found, uploading from local registry")
```

Comment:
https://discourse.ubuntu.com/t/punctuation/18884

Reply:
and in other places

Reply:
Improved here and in other parts.